src/uproot/models/RNTuple.py

```diff
@@ -218,26 +217,56 @@ def page_list_envelopes(self):
 
         return self.column_innerlist_dict
 
-    def col_form(self, field_id):
-        for cr in self.header.column_records:
+    def base_col_form(self, cr, col_id, parameters=None):
+        form_key = f"column-{col_id}"
+        if cr.type == uproot.const.rt_role_union:  # switch (UnionForm)
+            return form_key
+        elif cr.type > uproot.const.rt_role_struct:  # data column
+            return ak._v2.forms.NumpyForm(
+                _rntuple_col_types[cr.type], form_key=form_key, parameters=parameters
+            )
+        else:  # offset index column
+            return form_key
+
+    def col_form(self, field_id, type_name=None):
+        # FIXME remove this ugly logic
+        rel_crs = []
+        rel_crs_idxs = []
+        for i, cr in enumerate(self.header.column_records):
             if cr.field_id == field_id:
-                form_key = f"field-{cr.field_id}"
-                if cr.type > 2:  # data column
-                    return ak._v2.forms.NumpyForm(
-                        _rntuple_col_types[cr.type], form_key=form_key
-                    )
-                else:  # offset index column
-                    return form_key
+                rel_crs.append(cr)
+                rel_crs_idxs.append(i)
+            if cr.field_id > field_id:
+                break
+        if len(rel_crs) == 1:  # base case
+            return self.base_col_form(
+                rel_crs[0], rel_crs_idxs[0]
+            )
+        elif type_name == "std::string":  # string field splits->2 in col records
+            assert len(rel_crs_idxs) == 2
+            cr_char = rel_crs[-1]
+            assert cr_char.type == 5  # char
```

Comment:
Named constant.

Reply:
in this case the `5` is in `_rntuple_col_types`, do we want to have a reverse look up table?

Reply:
Reverse lookup tables can be useful and they can be built at startup, like this `_primitive_to_dtype_dict` and `_dtype_to_primitive_dict` pair:

https://github.com/scikit-hep/awkward/blob/af0f4569ed90eab64690098e5a64ae610282c4bf/src/awkward/_v2/types/numpytype.py#L66-L93

If that's what's needed here, go ahead and make the reverse lookup.

Reply:
(In my example, the use of the word "dict" in the variable name is because there are also functions:

https://github.com/scikit-hep/awkward/blob/af0f4569ed90eab64690098e5a64ae610282c4bf/src/awkward/_v2/types/numpytype.py#L23-L56

That part didn't need to be copied into `rntuple_col_num_to_dtype_dict` and `rntuple_col_type_to_num_dict`, but it's fine either way.)